File: sdk-go/client_test.go
package rbac

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func testClient(t *testing.T, h http.HandlerFunc, opts ...Option) (*Client, *httptest.Server) {
	t.Helper()
	srv := httptest.NewServer(h)
	t.Cleanup(srv.Close)
	c, err := NewClient(srv.URL, opts...)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return c, srv
}

func writeResp(w http.ResponseWriter, status int, body string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_, _ = io.WriteString(w, body)
}

// recorded captures the shape of one inbound request for assertions.
type recorded struct {
	method string
	path   string
	query  map[string][]string
	body   []byte
	header http.Header
}

func recorder(status int, body string, rec *recorded) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		b, _ := io.ReadAll(r.Body)
		rec.method = r.Method
		rec.path = r.URL.Path
		rec.query = r.URL.Query()
		rec.body = b
		rec.header = r.Header.Clone()
		writeResp(w, status, body)
	}
}

func TestNewClientValidation(t *testing.T) {
	tests := []struct {
		name    string
		baseURL string
		opts    []Option
		wantErr bool
	}{
		{"http ok", "http://localhost:8080", nil, false},
		{"https ok", "https://localhost:8443", nil, false},
		{"empty", "", nil, true},
		{"bad scheme", "ftp://localhost", nil, true},
		{"no host", "http://", nil, true},
		{"nil http client", "http://localhost", []Option{WithHTTPClient(nil)}, true},
		{"empty ca pem", "https://localhost", []Option{WithCACert(nil)}, true},
		{"bad ca pem", "https://localhost", []Option{WithCACert([]byte("not-pem"))}, true},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			_, err := NewClient(tc.baseURL, tc.opts...)
			if (err != nil) != tc.wantErr {
				t.Fatalf("NewClient(%q) err=%v wantErr=%v", tc.baseURL, err, tc.wantErr)
			}
		})
	}
}

func TestHealth(t *testing.T) {
	t.Run("ok", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			writeResp(w, http.StatusOK, `{"status":"ok"}`)
		})
		if err := c.Health(context.Background()); err != nil {
			t.Fatalf("Health: %v", err)
		}
	})
	t.Run("paused", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			writeResp(w, http.StatusServiceUnavailable, `{"status":"paused","reason":"disk mismatch"}`)
		})
		err := c.Health(context.Background())
		if !IsPaused(err) {
			t.Fatalf("want paused, got %v", err)
		}
		var ae *APIError
		if !errors.As(err, &ae) || ae.Message != "disk mismatch" {
			t.Fatalf("want reason as message, got %v", err)
		}
	})
}

func TestEnforce(t *testing.T) {
	t.Run("allow with options", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusOK, `{"allow":true}`, &rec))
		now := time.Date(2026, 6, 15, 0, 0, 0, 0, time.UTC)
		ok, err := c.Enforce(context.Background(), "alice", "doc:42",
			WithScenarios([]string{"VIP", "EU"}), WithNow(now))
		if err != nil {
			t.Fatalf("Enforce: %v", err)
		}
		if !ok {
			t.Fatal("want allow true")
		}
		if rec.method != http.MethodPost || rec.path != "/v1/enforce" {
			t.Fatalf("bad request shape: %s %s", rec.method, rec.path)
		}
		var body struct {
			Subject   string    `json:"subject"`
			Target    string    `json:"target"`
			Scenarios []string  `json:"scenarios"`
			Now       time.Time `json:"now"`
		}
		if err := json.Unmarshal(rec.body, &body); err != nil {
			t.Fatalf("decode sent body: %v", err)
		}
		if body.Subject != "alice" || body.Target != "doc:42" {
			t.Fatalf("bad subject/target: %+v", body)
		}
		if len(body.Scenarios) != 2 || body.Scenarios[0] != "VIP" {
			t.Fatalf("bad scenarios: %+v", body.Scenarios)
		}
		if !body.Now.Equal(now) {
			t.Fatalf("bad now: %v", body.Now)
		}
	})
	t.Run("deny omits options", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusOK, `{"allow":false}`, &rec))
		ok, err := c.Enforce(context.Background(), "a", "b")
		if err != nil || ok {
			t.Fatalf("want deny, got ok=%v err=%v", ok, err)
		}
		var raw map[string]json.RawMessage
		_ = json.Unmarshal(rec.body, &raw)
		if _, present := raw["scenarios"]; present {
			t.Fatal("scenarios should be omitted when empty")
		}
		if _, present := raw["now"]; present {
			t.Fatal("now should be omitted when unset")
		}
	})
}

func TestReachable(t *testing.T) {
	var rec recorded
	c, _ := testClient(t, recorder(http.StatusOK,
		`{"subject":"alice","reachable":["alice","doc:42"]}`, &rec))
	got, err := c.Reachable(context.Background(), "alice", WithScenarios([]string{"VIP", "EU"}))
	if err != nil {
		t.Fatalf("Reachable: %v", err)
	}
	if len(got) != 2 || got[0] != "alice" || got[1] != "doc:42" {
		t.Fatalf("bad result: %v", got)
	}
	if rec.query["subject"][0] != "alice" {
		t.Fatalf("bad subject query: %v", rec.query)
	}
	if sc := rec.query["scenario"]; len(sc) != 2 || sc[0] != "VIP" || sc[1] != "EU" {
		t.Fatalf("bad scenario query: %v", sc)
	}
}

func TestBindingCRUD(t *testing.T) {
	ctx := context.Background()
	sample := `{"src":"alice","dst":"role:editor","scenario":"","enabled":true,"conditions":[{"kind":"ALL"}]}`

	t.Run("ListBindings", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			writeResp(w, http.StatusOK, `{"bindings":[`+sample+`]}`)
		})
		bs, err := c.ListBindings(ctx)
		if err != nil {
			t.Fatalf("ListBindings: %v", err)
		}
		if len(bs) != 1 || bs[0].Src != "alice" {
			t.Fatalf("bad list: %+v", bs)
		}
	})

	t.Run("GetBinding ok", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusOK, sample, &rec))
		b, err := c.GetBinding(ctx, "alice", "role:editor", "")
		if err != nil {
			t.Fatalf("GetBinding: %v", err)
		}
		if b.Dst != "role:editor" {
			t.Fatalf("bad binding: %+v", b)
		}
		if rec.query["src"][0] != "alice" || rec.query["dst"][0] != "role:editor" {
			t.Fatalf("bad query: %v", rec.query)
		}
	})

	t.Run("GetBinding not found", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			writeResp(w, http.StatusNotFound, `{"error":"binding not found"}`)
		})
		_, err := c.GetBinding(ctx, "x", "y", "")
		if !IsNotFound(err) {
			t.Fatalf("want not found, got %v", err)
		}
	})

	t.Run("AddBinding created", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusCreated, sample, &rec))
		b, err := c.AddBinding(ctx, Binding{
			Src: "alice", Dst: "role:editor", Enabled: true,
			Conditions: []Condition{AllCondition()},
		})
		if err != nil {
			t.Fatalf("AddBinding: %v", err)
		}
		if b.Src != "alice" {
			t.Fatalf("bad echo: %+v", b)
		}
		if rec.method != http.MethodPost {
			t.Fatalf("want POST, got %s", rec.method)
		}
		var sent Binding
		if err := json.Unmarshal(rec.body, &sent); err != nil {
			t.Fatalf("decode sent: %v", err)
		}
		if len(sent.Conditions) != 1 || sent.Conditions[0].Kind != KindAll {
			t.Fatalf("bad sent conditions: %+v", sent.Conditions)
		}
	})

	t.Run("AddBinding conflict", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			writeResp(w, http.StatusConflict, `{"error":"binding (src,dst,scenario) exists"}`)
		})
		_, err := c.AddBinding(ctx, Binding{Src: "a", Dst: "b"})
		if !IsConflict(err) {
			t.Fatalf("want conflict, got %v", err)
		}
	})

	t.Run("UpdateBinding ok", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusOK, sample, &rec))
		if _, err := c.UpdateBinding(ctx, Binding{Src: "alice", Dst: "role:editor", Enabled: true}); err != nil {
			t.Fatalf("UpdateBinding: %v", err)
		}
		if rec.method != http.MethodPut {
			t.Fatalf("want PUT, got %s", rec.method)
		}
	})

	t.Run("SetEnabled", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, recorder(http.StatusOK, `{"ok":true}`, &rec))
		if err := c.SetEnabled(ctx, "alice", "role:editor", "", false); err != nil {
			t.Fatalf("SetEnabled: %v", err)
		}
		if rec.method != http.MethodPatch || rec.path != "/v1/bindings/enabled" {
			t.Fatalf("bad shape: %s %s", rec.method, rec.path)
		}
		var body struct {
			Enabled bool `json:"enabled"`
		}
		_ = json.Unmarshal(rec.body, &body)
		if body.Enabled {
			t.Fatal("want enabled false in body")
		}
	})

	t.Run("RemoveBinding no content", func(t *testing.T) {
		var rec recorded
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			rec.method = r.Method
			rec.path = r.URL.Path
			rec.query = r.URL.Query()
			w.WriteHeader(http.StatusNoContent)
		})
		if err := c.RemoveBinding(ctx, "alice", "role:editor", "VIP"); err != nil {
			t.Fatalf("RemoveBinding: %v", err)
		}
		if rec.method != http.MethodDelete {
			t.Fatalf("want DELETE, got %s", rec.method)
		}
		if rec.query["scenario"][0] != "VIP" {
			t.Fatalf("bad scenario query: %v", rec.query)
		}
	})
}

func TestErrorMapping(t *testing.T) {
	tests := []struct {
		name   string
		status int
		body   string
		check  func(error) bool
		msg    string
	}{
		{"bad request", http.StatusBadRequest, `{"error":"resource id contains reserved char"}`, IsBadRequest, "resource id contains reserved char"},
		{"not found", http.StatusNotFound, `{"error":"binding not found"}`, IsNotFound, "binding not found"},
		{"conflict", http.StatusConflict, `{"error":"binding (src,dst,scenario) exists"}`, IsConflict, "binding (src,dst,scenario) exists"},
		{"paused", http.StatusServiceUnavailable, `{"error":"service paused","reason":"x"}`, IsPaused, "service paused"},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
				writeResp(w, tc.status, tc.body)
			})
			err := c.Health(context.Background())
			if !tc.check(err) {
				t.Fatalf("predicate failed for %v", err)
			}
			var ae *APIError
			if !errors.As(err, &ae) {
				t.Fatalf("want *APIError, got %T", err)
			}
			if ae.Message != tc.msg || ae.StatusCode != tc.status {
				t.Fatalf("bad APIError: %+v", ae)
			}
		})
	}

	t.Run("non-json body preserved", func(t *testing.T) {
		c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusInternalServerError)
			_, _ = io.WriteString(w, "boom")
		})
		err := c.Health(context.Background())
		var ae *APIError
		if !errors.As(err, &ae) {
			t.Fatalf("want *APIError, got %v", err)
		}
		if ae.Message != "" || string(ae.Body) != "boom" {
			t.Fatalf("bad APIError: %+v", ae)
		}
	})
}

func TestRequestHeaders(t *testing.T) {
	var rec recorded
	c, _ := testClient(t, recorder(http.StatusOK, `{"allow":true}`, &rec), WithUserAgent("custom/1.0"))
	if _, err := c.Enforce(context.Background(), "a", "b"); err != nil {
		t.Fatalf("Enforce: %v", err)
	}
	if got := rec.header.Get("User-Agent"); got != "custom/1.0" {
		t.Fatalf("bad User-Agent: %q", got)
	}
	if got := rec.header.Get("Accept"); got != "application/json" {
		t.Fatalf("bad Accept: %q", got)
	}
	if got := rec.header.Get("Content-Type"); got != "application/json" {
		t.Fatalf("bad Content-Type: %q", got)
	}
}

func TestContextCancellation(t *testing.T) {
	c, _ := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeResp(w, http.StatusOK, `{"allow":true}`)
	})
	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	if _, err := c.Enforce(ctx, "a", "b"); err == nil {
		t.Fatal("want error from cancelled context")
	}
}

File: sdk-rust/tests/client_test.rs
//! Client tests mirroring sdk-go coverage (HTTP via mockito).

use std::time::Duration;

use chrono::{TimeZone, Utc};
use serde_json::Value;

use rbac::{
    all_condition, Binding, CallOpts, Client, ConditionKind, ErrorKind,
};

fn client_for(server: &mockito::ServerGuard) -> Client {
    Client::new(&server.url()).expect("Client::new")
}

#[test]
fn new_client_validation() {
    assert!(Client::new("http://localhost:8080").is_ok());
    assert!(Client::new("https://localhost:8443").is_ok());
    assert!(Client::new("").is_err());
    assert!(Client::new("ftp://localhost").is_err());
    assert!(Client::new("http://").is_err());

    let err = Client::builder("https://localhost")
        .unwrap()
        .ca_cert(&[] as &[u8])
        .unwrap_err();
    assert_eq!(err.kind(), ErrorKind::Config);

    let err = Client::builder("https://localhost")
        .unwrap()
        .ca_cert(b"not-pem")
        .unwrap()
        .build()
        .unwrap_err();
    assert_eq!(err.kind(), ErrorKind::Config);

    let err = Client::builder("http://localhost")
        .unwrap()
        .insecure_skip_verify(true)
        .build()
        .unwrap_err();
    assert!(err.to_string().contains("http://"));
}

#[test]
fn health_ok_and_paused() {
    let mut server = mockito::Server::new();
    let _m = server
        .mock("GET", "/healthz")
        .with_status(200)
        .with_body(r#"{"status":"ok"}"#)
        .create();
    client_for(&server).health().unwrap();

    let mut server = mockito::Server::new();
    let _m = server
        .mock("GET", "/healthz")
        .with_status(503)
        .with_body(r#"{"status":"paused","reason":"disk mismatch"}"#)
        .create();
    let err = client_for(&server).health().unwrap_err();
    assert!(err.is_paused());
    match err {
        rbac::Error::Api(ae) => assert_eq!(ae.message, "disk mismatch"),
        other => panic!("want ApiError, got {other}"),
    }
}

#[test]
fn enforce_with_options() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("POST", "/v1/enforce")
        .match_header("content-type", "application/json")
        .with_status(200)
        .with_body(r#"{"allow":true}"#)
        .create();

    let now = Utc.with_ymd_and_hms(2026, 6, 15, 0, 0, 0).unwrap();
    let ok = client_for(&server)
        .enforce(
            "alice",
            "doc:42",
            CallOpts::new()
                .with_scenarios(["VIP", "EU"])
                .with_now(now),
        )
        .unwrap();
    assert!(ok);
    m.assert();
}

#[test]
fn enforce_omits_empty_options() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("POST", "/v1/enforce")
        .with_status(200)
        .with_body(r#"{"allow":false}"#)
        .match_body(mockito::Matcher::Regex(
            r#"\{"subject":"a","target":"b"\}"#.into(),
        ))
        .create();

    let ok = client_for(&server)
        .enforce("a", "b", CallOpts::default())
        .unwrap();
    assert!(!ok);
    m.assert();
}

#[test]
fn reachable_query() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/v1/reachable")
        .match_query(mockito::Matcher::Exact(
            "subject=alice&scenario=VIP&scenario=EU".into(),
        ))
        .with_status(200)
        .with_body(r#"{"subject":"alice","reachable":["alice","doc:42"]}"#)
        .create();

    let got = client_for(&server)
        .reachable("alice", CallOpts::new().with_scenarios(["VIP", "EU"]))
        .unwrap();
    assert_eq!(got, vec!["alice", "doc:42"]);
    m.assert();
}

#[test]
fn binding_crud() {
    let sample = r#"{"src":"alice","dst":"role:editor","scenario":"","enabled":true,"conditions":[{"kind":"ALL"}]}"#;

    let mut server = mockito::Server::new();
    let _m = server
        .mock("GET", "/v1/bindings")
        .match_query(mockito::Matcher::Missing)
        .with_status(200)
        .with_body(format!(r#"{{"bindings":[{sample}]}}"#))
        .create();
    let bs = client_for(&server).list_bindings().unwrap();
    assert_eq!(bs.len(), 1);
    assert_eq!(bs[0].src, "alice");
    assert_eq!(bs[0].conditions[0].kind, ConditionKind::All);

    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/v1/bindings")
        .match_query(mockito::Matcher::AllOf(vec![
            mockito::Matcher::UrlEncoded("src".into(), "alice".into()),
            mockito::Matcher::UrlEncoded("dst".into(), "role:editor".into()),
        ]))
        .with_status(200)
        .with_body(sample)
        .create();
    let b = client_for(&server)
        .get_binding("alice", "role:editor", "")
        .unwrap();
    assert_eq!(b.dst, "role:editor");
    m.assert();

    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/v1/bindings")
        .match_query(mockito::Matcher::Regex(r"src=x&dst=y".into()))
        .with_status(404)
        .with_body(r#"{"error":"binding not found"}"#)
        .create();
    let err = client_for(&server)
        .get_binding("x", "y", "")
        .unwrap_err();
    assert!(err.is_not_found(), "got {err:?}");
    m.assert();

    let mut server = mockito::Server::new();
    let m = server
        .mock("POST", "/v1/bindings")
        .with_status(200)
        .with_body(sample)
        .create();
    let b = Binding::new("alice", "role:editor")
        .with_enabled(true)
        .with_conditions(vec![all_condition()]);
    let out = client_for(&server).add_binding(&b).unwrap();
    assert_eq!(out.src, "alice");
    m.assert();

    let mut server = mockito::Server::new();
    let _m = server
        .mock("POST", "/v1/bindings")
        .with_status(409)
        .with_body(r#"{"error":"duplicate"}"#)
        .create();
    let err = client_for(&server).add_binding(&b).unwrap_err();
    assert!(err.is_conflict());

    let mut server = mockito::Server::new();
    let m = server
        .mock("PUT", "/v1/bindings")
        .with_status(200)
        .with_body(sample)
        .create();
    client_for(&server).update_binding(&b).unwrap();
    m.assert();

    let mut server = mockito::Server::new();
    let m = server
        .mock("PATCH", "/v1/bindings/enabled")
        .match_body(mockito::Matcher::Json(serde_json::json!({
            "src": "alice",
            "dst": "role:editor",
            "scenario": "",
            "enabled": false
        })))
        .with_status(200)
        .with_body("{}")
        .create();
    client_for(&server)
        .set_enabled("alice", "role:editor", "", false)
        .unwrap();
    m.assert();

    let mut server = mockito::Server::new();
    let m = server
        .mock("DELETE", "/v1/bindings")
        .match_query(mockito::Matcher::AllOf(vec![
            mockito::Matcher::UrlEncoded("src".into(), "alice".into()),
            mockito::Matcher::UrlEncoded("dst".into(), "role:editor".into()),
            mockito::Matcher::UrlEncoded("scenario".into(), "VIP".into()),
        ]))
        .with_status(200)
        .with_body("")
        .create();
    client_for(&server)
        .remove_binding("alice", "role:editor", "VIP")
        .unwrap();
    m.assert();
}

#[test]
fn user_agent_and_timeout_options() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/healthz")
        .match_header("user-agent", "my-agent/1")
        .with_status(200)
        .with_body("{}")
        .create();
    let c = Client::builder(&server.url())
        .unwrap()
        .user_agent("my-agent/1")
        .timeout(Duration::from_secs(5))
        .build()
        .unwrap();
    c.health().unwrap();
    m.assert();
}

#[test]
fn base_path_is_preserved() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/prefix/healthz")
        .with_status(200)
        .with_body("{}")
        .create();
    let base = format!("{}/prefix", server.url());
    Client::new(&base).unwrap().health().unwrap();
    m.assert();
}

#[test]
fn bad_request_kind() {
    let mut server = mockito::Server::new();
    let _m = server
        .mock("POST", "/v1/enforce")
        .with_status(400)
        .with_body(r#"{"error":"missing subject"}"#)
        .create();
    let err = client_for(&server)
        .enforce("a", "b", CallOpts::default())
        .unwrap_err();
    assert!(err.is_bad_request());
    assert_eq!(err.kind(), ErrorKind::BadRequest);
}

#[test]
fn ca_cert_file_missing() {
    let err = Client::builder("https://localhost:8443")
        .unwrap()
        .ca_cert_file("/nonexistent/ca.crt")
        .unwrap_err();
    assert_eq!(err.kind(), ErrorKind::Config);
}

#[test]
fn enforce_sends_correct_json_shape() {
    let now = Utc.with_ymd_and_hms(2026, 6, 15, 0, 0, 0).unwrap();
    let mut server = mockito::Server::new();
    let expected = serde_json::json!({
        "subject": "alice",
        "target": "doc:42",
        "scenarios": ["VIP"],
        "now": "2026-06-15T00:00:00Z"
    });
    let m = server
        .mock("POST", "/v1/enforce")
        .match_body(mockito::Matcher::Json(expected))
        .with_status(200)
        .with_body(r#"{"allow":true}"#)
        .create();
    client_for(&server)
        .enforce(
            "alice",
            "doc:42",
            CallOpts::new().with_scenarios(["VIP"]).with_now(now),
        )
        .unwrap();
    m.assert();
}

#[test]
fn list_bindings_request_has_no_query() {
    let mut server = mockito::Server::new();
    let m = server
        .mock("GET", "/v1/bindings")
        .match_query(mockito::Matcher::Missing)
        .with_status(200)
        .with_body(r#"{"bindings":[]}"#)
        .create();
    let v: Value = serde_json::from_str(r#"{"bindings":[]}"#).unwrap();
    assert!(v["bindings"].as_array().unwrap().is_empty());
    client_for(&server).list_bindings().unwrap();
    m.assert();
}
